Refactor model dates

dhjacobson opened the pull request.

dhjacobson committed:
make CovidModel inherit CovidModelSpecifications

dhjacobson committed:
fix issue where end date was not getting updated

dhjacobson committed:
Merge remote-tracking branch 'origin/develop' into refactor-model-dates

fouticus commented on covid_model/model_specs.py:
```diff
@@ -15,10 +15,10 @@
 
         self.start_date = start_date
-        self.end_date = end_date
+        self.end_date = None
```
a little confusing to see different initializations for start_date and end_date. I see on lines 53 and 71 the end_date argument is finally used, but setting `self.end_date = end_date` here would make it clear that it's the default. Then lines 53 and 71 could check if self.end_date is none

dhjacobson replied:
Good point, will change. As a high-level, there is an important difference between the start_date and the end_date in the specs: end_date can be changed (using update specs) and start_date cannot. Your point is still correct; just thought I'd highlight that.

fouticus commented on covid_model/model_specs.py:
```diff
@@ -37,85 +37,104 @@ def __init__(self, start_date=dt.date(2020, 1, 24), end_date=dt.date(2022, 5, 31
+                raise TypeError(f'from_specs must be an int or CovidModelSpecifications; not a {type(from_specs)}.')
+
+        if start_date is not None:
+            if self.start_date is not None and start_date != self.start_date:
```
if you adjust based on my comments above, then I think you just have to check `start_date != self.start_date` here.

dhjacobson replied:
I'm actually going to go the other way with these two comments, and not set either self.start_date or self.end_date at the beginning. That way it's clear that what we're doing is taking the dates from from_specs (if provided) and overwriting them (if possible). Does that make sense?

fouticus commented on covid_model/model.py:
```diff
@@ -240,11 +230,11 @@ def immunity(self, variant='omicron', vacc_only=False, to_hosp=False):
             return (n * variant_params['immunity']).groupby('t').sum() / n.groupby('t').sum()
 
     # write to covid_model.results in Postgres
```
stale comment

fouticus commented on covid_model/model.py:
```diff
@@ -240,11 +230,11 @@ def immunity(self, variant='omicron', vacc_only=False, to_hosp=False):
 
     # write to covid_model.results in Postgres
-    def write_to_db(self, engine=None, new_spec=False, vals_json_attr='seir', cmpts_json_attrs=('age', 'vacc'), sim_id=None, sim_result_id=None):
+    def write_results_to_db(self, engine=None, new_spec=False, vals_json_attr='seir', cmpts_json_attrs=('age', 'vacc'), sim_id=None, sim_result_id=None):
```
rename to prevent overloading I assume

dhjacobson replied:
Yep, exactly. This was the only function or attribute that had that issue --- except for of course the dates, which is the whole point of this.

fouticus commented on covid_model/run_model_scenarios.py:
```diff
@@ -113,19 +113,19 @@ def main():
         model.prep(specs=specs)
         print(f'Running scenarios...')
-        model.specifications.tags.update({'run_type': 'Vaccination Scenario', 'vacc_cap': vacc_scen, 'tc_shift': 'no shift', 'tc_shift_date': 'no_shift'})
+        model.tags.update_specs({'run_type': 'Vaccination Scenario', 'vacc_cap': vacc_scen, 'tc_shift': 'no shift', 'tc_shift_date': 'no_shift'})
```
`CovidModelSpecifications.tags` is a dict, right? should this be `model.update_specs`?

fouticus replied:
what am I missing?

dhjacobson replied:
Whoops, this was a mistake, good catch. PyCharm refactor mistake. This script is kind of messed up for other reasons, so I didn't test it.

fouticus commented on covid_model/model_fit.py:
```diff
@@ -7,15 +7,15 @@
-from data_imports import ExternalHosps, ExternalData
-from model import CovidModel
-from model_specs import CovidModelSpecifications
+from covid_model.data_imports import ExternalHosps, ExternalData
```
this avoids having covid_model in the PYTHONPATH, right?

dhjacobson replied:
Yes, I think so. I'd fixed it everywhere else already, but if you see any others we should update everything to use `covid_model.`

fouticus commented on covid_model/model_fit.py:
```diff
@@ -97,13 +97,13 @@ def run(self, engine, method='curve_fit', window_size=14, look_back=None,
         self.fitted_tc_cov = fitted_tc_cov
         self.fitted_model = model
-        self.fitted_specs = model.specifications
+        self.fitted_specs = model
```
this messes with me. as long as `self.fitted_specs` is always treated as an instance of `CovidModelSpecifications` and never treated as an instance of `CovidModel` then this seems fine. But it relies on the programmer's obeying that constraint. Seems like a necessary consequence of duck typing. Or am I way off base?

dhjacobson replied:
It looks to me like we can just get rid of fitted_specs, since it doesn't look like it's in use anywhere. I was just moving fast and replacing stuff --- but you're right that in this case the result is silly.

fouticus reviewed (commented):
Looks good. a few questions and comments that you can take or leave. One question about model.tags.update_specs that I'm probably just not following, but let me know what you think. And also see my comment in model_fit.py, am I understanding correctly what you're doing?

dhjacobson committed:
don't set specs start_date twice

dhjacobson committed:
fix stale comment

dhjacobson committed:
fix update --> update_specs refactor bug

dhjacobson committed:
remove superfluous fit.fitted_specs

dhjacobson commented:
Thanks for the review! Adjusted the start/end date setting per your feedback, fixed stale comment, fixed the update_specs typo, and trashed fitted_specs because it wasn't doing anything.
